File: Sources/Bifrost/Services/SagaStats.swift
import Foundation

/// Builds a read-only "Saga" snapshot for Home's stats card: total Valheim
/// playtime (from Steam's `localconfig.vdf`), installed mod count and
/// multiplayer-safety breakdown, backup count/size, and world/character
/// names — all presented with a bit of viking flair.
///
/// Deliberately takes every input as a plain value rather than reaching out
/// and fetching things itself (manifest, index, backups, save directory,
/// localconfig text) — that keeps this trivially testable against fixtures
/// (see `DebugCheck`'s "fun" section) and lets `SagaCard` reuse state the
/// rest of the app has already loaded instead of re-fetching it. The only
/// I/O this file performs directly is a read-only directory listing of the
/// save directory to discover world/character file names and sizes.
enum SagaStats {
    /// One multiplayer-safety class's share of the installed mods (see
    /// `ModClassifier`).
    struct ClassCount: Sendable, Equatable {
        let modClass: ModClass
        let count: Int
    }

    /// One world or character save file (or pair of files, for a world),
    /// grouped by its name stem.
    struct SaveEntry: Sendable, Equatable, Identifiable {
        let name: String
        let byteSize: Int64
        var id: String { name }
    }

    struct Snapshot: Sendable, Equatable {
        /// `nil` when `localconfig.vdf` couldn't be read or had no
        /// `Playtime` entry for Valheim yet (a fresh Steam library, or the
        /// game has never been launched through Steam).
        var playtimeMinutes: Int?
        var modCount: Int
        /// Only classes with at least one installed mod, in `ModClass`'s
        /// declared order.
        var classBreakdown: [ClassCount]
        var backupCount: Int
        var backupTotalBytes: Int64
        var worlds: [SaveEntry]
        var characters: [SaveEntry]

        static let empty = Snapshot(playtimeMinutes: nil, modCount: 0, classBreakdown: [], backupCount: 0, backupTotalBytes: 0, worlds: [], characters: [])
    }

    /// Builds a snapshot from already-loaded pieces. `localConfigText` is
    /// the raw contents of the most-recently-used Steam profile's
    /// `localconfig.vdf` (see `SteamConfigurator.realLocalConfigURL`), or
    /// `nil` if it couldn't be read — every field degrades gracefully when
    /// its source is missing rather than throwing.
    static func snapshot(
        manifest: InstalledManifest,
        index: [ThunderstorePackage],
        backups: [SaveBackup.Backup],
        saveDir: URL,
        localConfigText: String?
    ) -> Snapshot {
        Snapshot(
            playtimeMinutes: localConfigText.flatMap(playtimeMinutes(inLocalConfig:)),
            modCount: manifest.mods.count,
            classBreakdown: classBreakdown(mods: manifest.mods, index: index),
            backupCount: backups.count,
            backupTotalBytes: backups.reduce(Int64(0)) { $0 + $1.byteSize },
            worlds: worldEntries(in: saveDir),
            characters: characterEntries(in: saveDir)
        )
    }

    /// Reads Valheim's (`892970`) `Playtime` key (minutes) from a
    /// `localconfig.vdf`'s text, via `VDF`/`SteamConfigurator.appPath` — the
    /// same splice path Bifrost already uses for `LaunchOptions`.
    static func playtimeMinutes(inLocalConfig text: String) -> Int? {
        VDF.value(forKey: "Playtime", atPath: SteamConfigurator.appPath, in: text).flatMap { Int($0) }
    }

    /// Classifies every installed mod (`ModClassifier`) and tallies counts
    /// per class, omitting classes with zero mods.
    static func classBreakdown(mods: [InstalledManifest.InstalledMod], index: [ThunderstorePackage]) -> [ClassCount] {
        var counts: [String: Int] = [:]
        for mod in mods {
            let classification = ModClassifier.classify(mod: mod, index: index)
            counts[classification.modClass.rawValue, default: 0] += 1
        }
        return ModClass.allCases.compactMap { modClass in
            guard let count = counts[modClass.rawValue], count > 0 else { return nil }
            return ClassCount(modClass: modClass, count: count)
        }
    }

    /// Every `.fwl`/`.db` pair under `worlds_local` (falling back to the
    /// legacy `worlds` directory), grouped by name stem with their combined
    /// size.
    static func worldEntries(in saveDir: URL) -> [SaveEntry] {
        entries(in: resolveDir(saveDir, primary: "worlds_local", legacy: "worlds"), matchingExtensions: ["fwl", "db"])
    }

    /// Every `.fch` under `characters_local` (falling back to the legacy
    /// `characters` directory), one entry per character.
    static func characterEntries(in saveDir: URL) -> [SaveEntry] {
        entries(in: resolveDir(saveDir, primary: "characters_local", legacy: "characters"), matchingExtensions: ["fch"])
    }

    private static func resolveDir(_ saveDir: URL, primary: String, legacy: String) -> URL {
        let primaryURL = saveDir.appendingPathComponent(primary)
        return FileManager.default.fileExists(atPath: primaryURL.path) ? primaryURL : saveDir.appendingPathComponent(legacy)
    }

    private static func entries(in dir: URL, matchingExtensions extensions: [String]) -> [SaveEntry] {
        guard let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: [.fileSizeKey]) else {
            return []
        }
        var sizeByStem: [String: Int64] = [:]
        for file in files {
            guard extensions.contains(file.pathExtension.lowercased()) else { continue }
            let stem = file.deletingPathExtension().lastPathComponent
            let size: Int = (try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
            sizeByStem[stem, default: 0] += Int64(size)
        }
        return sizeByStem.map { SaveEntry(name: $0.key, byteSize: $0.value) }.sorted { $0.name < $1.name }
    }

    /// Short, viking-flavored one-line summaries of `snapshot`, in a fixed
    /// priority order (playtime, worlds, characters, mods, backups). Always
    /// returns at least one line, even when every source is missing.
    static func flavorLines(for snapshot: Snapshot) -> [String] {
        var lines: [String] = []

        if let minutes = snapshot.playtimeMinutes {
            let hours = minutes / 60
            lines.append("⚔️ \(hours) hour\(hours == 1 ? "" : "s") carved into the saga so far")
        }
        if !snapshot.worlds.isEmpty {
            lines.append("🌍 \(snapshot.worlds.count) world\(snapshot.worlds.count == 1 ? "" : "s") under your protection")
        }
        if !snapshot.characters.isEmpty {
            lines.append("🛡️ \(snapshot.characters.count) hero\(snapshot.characters.count == 1 ? "" : "es") answering the call")
        }
        if snapshot.modCount > 0 {
            lines.append("🔨 \(snapshot.modCount) mod\(snapshot.modCount == 1 ? "" : "s") reinforcing the longship")
        }
        if snapshot.backupCount > 0 {
            lines.append("💾 \(snapshot.backupCount) backup\(snapshot.backupCount == 1 ? "" : "s") guarding \(formatBytes(snapshot.backupTotalBytes)) of saga")
        }

        if lines.isEmpty {
            lines.append("🪓 No saga recorded yet — sharpen your axe and begin")
        }
        return lines
    }

    static func formatBytes(_ bytes: Int64) -> String {
        ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
    }
}

File: src/Bifrost.Core/Services/SagaStats.cs
using Bifrost.Core.Models;

namespace Bifrost.Core.Services;

/// <summary>
/// Builds the Home tab's "Saga" stats card: playtime (from Steam's
/// <c>localconfig.vdf</c>), world/character counts (from the Valheim save
/// directory), installed mods by risk class (via <see cref="ModClassifier"/>),
/// and backup count/size (via <see cref="SaveBackup"/>) — rendered as a
/// short list of viking-flavored one-liners. Ported from the macOS reference
/// implementation's <c>SagaStats.swift</c>.
/// </summary>
public static class SagaStats
{
    public sealed record ClassCount(ModClass ModClass, int Count);

    public sealed record SaveEntry(string Name, long ByteSize);

    public sealed class Snapshot
    {
        /// <summary>Null if localconfig.vdf is unreadable or has no Playtime key yet.</summary>
        public int? PlaytimeMinutes { get; init; }
        public int ModCount { get; init; }
        /// <summary>Only classes with count &gt; 0, in <see cref="ModClass"/> declaration order.</summary>
        public List<ClassCount> ClassBreakdown { get; init; } = new();
        public int BackupCount { get; init; }
        public long BackupTotalBytes { get; init; }
        public List<SaveEntry> Worlds { get; init; } = new();
        public List<SaveEntry> Characters { get; init; } = new();

        public static Snapshot Empty => new();
    }

    private static readonly string[] PlaytimePath = { "UserLocalConfigStore", "Software", "Valve", "Steam", "apps", GameLocator.ValheimAppId };

    /// <summary>
    /// Reads Valheim's Playtime (in minutes) out of a
    /// <c>localconfig.vdf</c>'s text, or null if unreadable/missing. See
    /// <see cref="VdfParser.FindNestedValue"/>.
    /// </summary>
    public static int? PlaytimeMinutes(string? localConfigText)
    {
        if (string.IsNullOrEmpty(localConfigText))
        {
            return null;
        }
        var value = VdfParser.FindNestedValue("Playtime", PlaytimePath, localConfigText);
        return value is not null && int.TryParse(value, out var minutes) ? minutes : null;
    }

    /// <summary>
    /// The most recently modified <c>userdata/&lt;profile&gt;/config/localconfig.vdf</c>
    /// under <paramref name="steamRoot"/> — mirrors picking "whichever
    /// numeric userdata profile has the most recent modification date",
    /// same convention the macOS app uses to pick a userdata profile without
    /// needing the user's actual SteamID. Null if no such file exists
    /// anywhere under <c>userdata</c>.
    /// </summary>
    public static string? FindMostRecentLocalConfig(string steamRoot)
    {
        var userdataDir = Path.Combine(steamRoot, "userdata");
        if (!Directory.Exists(userdataDir))
        {
            return null;
        }

        string? best = null;
        var bestTime = DateTime.MinValue;
        foreach (var profileDir in Directory.EnumerateDirectories(userdataDir))
        {
            var path = Path.Combine(profileDir, "config", "localconfig.vdf");
            if (!File.Exists(path))
            {
                continue;
            }
            var modified = File.GetLastWriteTimeUtc(path);
            if (best is null || modified > bestTime)
            {
                best = path;
                bestTime = modified;
            }
        }
        return best;
    }

    /// <summary>Tallies every installed mod's <see cref="ModClassifier"/> classification, keeping only non-empty classes in enum declaration order.</summary>
    public static List<ClassCount> ClassBreakdown(InstalledManifest manifest, IReadOnlyList<ThunderstorePackage> index)
    {
        var counts = new Dictionary<ModClass, int>();
        foreach (var mod in manifest.Mods)
        {
            var classification = ModClassifier.Classify(mod, index);
            counts[classification.ModClass] = counts.GetValueOrDefault(classification.ModClass) + 1;
        }

        var result = new List<ClassCount>();
        foreach (ModClass modClass in Enum.GetValues<ModClass>())
        {
            if (counts.TryGetValue(modClass, out var count) && count > 0)
            {
                result.Add(new ClassCount(modClass, count));
            }
        }
        return result;
    }

    /// <summary>Resolves <c>saveDir/worlds_local</c> (falling back to legacy <c>worlds</c>), groups <c>.fwl</c>/<c>.db</c> files by stem, sums each stem's file sizes.</summary>
    public static List<SaveEntry> WorldEntries(string saveDir) =>
        GroupedEntries(saveDir, "worlds_local", "worlds", new[] { ".fwl", ".db" });

    /// <summary>Resolves <c>saveDir/characters_local</c> (falling back to legacy <c>characters</c>), one entry per <c>.fch</c> file's stem.</summary>
    public static List<SaveEntry> CharacterEntries(string saveDir) =>
        GroupedEntries(saveDir, "characters_local", "characters", new[] { ".fch" });

    private static List<SaveEntry> GroupedEntries(string saveDir, string preferredSubdir, string legacySubdir, string[] extensions)
    {
        var dir = Directory.Exists(Path.Combine(saveDir, preferredSubdir))
            ? Path.Combine(saveDir, preferredSubdir)
            : Path.Combine(saveDir, legacySubdir);
        if (!Directory.Exists(dir))
        {
            return new List<SaveEntry>();
        }

        var sizesByStem = new Dictionary<string, long>(StringComparer.OrdinalIgnoreCase);
        try
        {
            foreach (var file in Directory.EnumerateFiles(dir))
            {
                var ext = Path.GetExtension(file);
                if (!extensions.Any(e => string.Equals(e, ext, StringComparison.OrdinalIgnoreCase)))
                {
                    continue;
                }
                var stem = Path.GetFileNameWithoutExtension(file);
                long size;
                try { size = new FileInfo(file).Length; } catch { size = 0; }
                sizesByStem[stem] = sizesByStem.GetValueOrDefault(stem) + size;
            }
        }
        catch
        {
            return new List<SaveEntry>();
        }

        return sizesByStem
            .OrderBy(kv => kv.Key, StringComparer.OrdinalIgnoreCase)
            .Select(kv => new SaveEntry(kv.Key, kv.Value))
            .ToList();
    }

    /// <summary>Builds a full snapshot from already-loaded state — no I/O of its own beyond the small helpers above, so callers gather manifest/index/backups/localConfigText themselves first.</summary>
    public static Snapshot BuildSnapshot(InstalledManifest manifest, IReadOnlyList<ThunderstorePackage> index, IReadOnlyList<SaveBackup.Backup> backups, string saveDir, string? localConfigText) => new()
    {
        PlaytimeMinutes = PlaytimeMinutes(localConfigText),
        ModCount = manifest.Mods.Count,
        ClassBreakdown = ClassBreakdown(manifest, index),
        BackupCount = backups.Count,
        BackupTotalBytes = backups.Sum(b => b.ByteSize),
        Worlds = WorldEntries(saveDir),
        Characters = CharacterEntries(saveDir),
    };

    /// <summary>
    /// A prioritized, variable-length list of viking-flavored one-liners —
    /// each only appended if its data is present/non-zero. Always returns at
    /// least 1 line (a fallback when everything's empty), at most 5.
    /// </summary>
    public static List<string> FlavorLines(Snapshot snapshot)
    {
        var lines = new List<string>();

        if (snapshot.PlaytimeMinutes is { } minutes)
        {
            var hours = minutes / 60;
            lines.Add($"⚔️ {hours} hour{(hours == 1 ? "" : "s")} carved into the saga so far");
        }
        if (snapshot.Worlds.Count > 0)
        {
            lines.Add($"🌍 {snapshot.Worlds.Count} world{(snapshot.Worlds.Count == 1 ? "" : "s")} under your protection");
        }
        if (snapshot.Characters.Count > 0)
        {
            lines.Add($"🛡️ {snapshot.Characters.Count} hero{(snapshot.Characters.Count == 1 ? "" : "es")} answering the call");
        }
        if (snapshot.ModCount > 0)
        {
            lines.Add($"🔨 {snapshot.ModCount} mod{(snapshot.ModCount == 1 ? "" : "s")} reinforcing the longship");
        }
        if (snapshot.BackupCount > 0)
        {
            lines.Add($"💾 {snapshot.BackupCount} backup{(snapshot.BackupCount == 1 ? "" : "s")} guarding {FormatBytes(snapshot.BackupTotalBytes)} of saga");
        }

        if (lines.Count == 0)
        {
            lines.Add("🪓 No saga recorded yet — sharpen your axe and begin");
        }
        return lines;
    }

    /// <summary>Decimal (1000-based) human-readable byte formatting, matching macOS's <c>ByteCountFormatter(.file)</c> convention (Finder-style "1.2 GB") rather than a binary 1024-based one.</summary>
    public static string FormatBytes(long bytes)
    {
        string[] units = { "bytes", "KB", "MB", "GB", "TB" };
        double value = bytes;
        var unitIndex = 0;
        while (value >= 1000 && unitIndex < units.Length - 1)
        {
            value /= 1000;
            unitIndex++;
        }
        return unitIndex == 0
            ? $"{bytes} bytes"
            : $"{value.ToString("0.#", System.Globalization.CultureInfo.InvariantCulture)} {units[unitIndex]}";
    }
}
